src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -600,11 +600,12 @@ public override void List(string query, QueryParameters queryParameters, IList r
 			{
 				CheckAndUpdateSessionStatus();
 				queryParameters.ValidateParameters();
-				var plan = GetHQLQueryPlan(query, false);
-				AutoFlushIfRequired(plan.QuerySpaces);
+				var plan = GetHQLQueryPlan(query, false);
+                
+                dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
+                AutoFlushIfRequired(plan.QuerySpaces);
 
 				bool success = false;
-				dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
 				try
```

Comment:
The try catch statement is also about ensuring `dontFlushFromFind` get decremented before leaving the method, whatever happened. Calling `AutoFlushIfRequired(plan.QuerySpaces);` between `dontFlushFromFind` and the try is a no-go.
A dedicated try finally looks now required for ensuring `dontFlushFromFind` decrement without trigerring  `AfterOperation` if the auto flushed failed.